influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,266 @@
+            pd.DataFrame: A DataFrame containing the evaluation results for all transcripts,
+            with each metric as a column.
+        """
+        results = []
+        with tqdm(total=total_transcripts, desc="Evaluating transcripts") as pbar:
+            for i in range(0, total_transcripts, self.batch_size):
+                batch = all_transcripts_with_env[i : i + self.batch_size]
+                batch_results = self.evaluate_batch(batch, i, None)
+                results.extend(batch_results)
+                pbar.update(len(batch))
+
+        return self.process_results(results)
+
+    def sync_evaluate_iteration(self, all_transcripts_with_env, total_transcripts):
+        """
+        Evaluate all transcripts for the current iteration using multiprocessing.
+
+        Args:
+            all_transcripts_with_env (List[Tuple[List[Dict], str]]): List of tuples containing transcripts and
+            their corresponding environment names.
+            total_transcripts (int): Total number of transcripts to evaluate.
+
+        Returns:
+            pd.DataFrame: A DataFrame containing evaluation results for all transcripts.
+        """
+        with multiprocessing.Pool(processes=len(self.devices)) as pool:
+            results = []
+            with tqdm(total=total_transcripts, desc="Evaluating transcripts") as pbar:
+                # Process transcripts in batches, where batch size = self.batch_size * number of devices
+                for i in range(0, total_transcripts, self.batch_size * len(self.devices)):
+                    device_batches = []
+
+                    # Create batches for each device
+                    for j, device in enumerate(self.devices):
+                        start = i + j * self.batch_size
+                        end = min(start + self.batch_size, total_transcripts)
+
+                        if start >= total_transcripts:
+                            break
+
+                        # Append a tuple containing the batch, start index, and device
+                        device_batches.append((all_transcripts_with_env[start:end], start, device))
+
+                    if not device_batches:
+                        break
+
+                    # Process batches in parallel using starmap
+                    batch_results = pool.starmap(self.evaluate_batch, device_batches)
+                    results.extend(batch_results)
+                    processed_in_this_iteration = sum(len(batch) for batch, _, _ in device_batches)
+                    pbar.update(processed_in_this_iteration)
+
+        flat_results = list(itertools.chain(*results))
+        return self.process_results(flat_results)
+
+    def evaluate_batch(self, batch, start_index, device):
+        backend = self.backend_class(
+            model_name=self.backend_config["model_name"],
+            model_id=self.backend_config["model_id"],
+            lora_path=self.backend_config["lora_path"],
+            device=device,
+        )
```

Comment:
This step takes some time to run – seems strange to have it repeated it every batch. Is there maybe a way to give multiple batches to a process and have the process go through multiple batches, so that we don't have to reload the backend a bunch of times?

Reply:
I see. The simplest way I can think of to avoid this issue is to just have a self.backends_lst in the constructor, with one element for each device. Then we can just load the backends once and pull it out in this function